File: Data Structures & Algorithms/insertionSort/submission-3.java
// Definition for a pair
// class Pair {
//     int key;
//     String value;
//
//     Pair(int key, String value) {
//         this.key = key;
//         this.value = value;
//     }
// }
public class Solution {
    public List<List<Pair>> insertionSort(List<Pair> pairs) {
        List<List<Pair>> arr = new ArrayList<>();
        if (pairs.isEmpty()) {
            return arr;
        }
        arr.add(new ArrayList<>(pairs));
        for (int i = 1; i<pairs.size(); i++) {
            int j=i-1;
            while (j>=0 && pairs.get(j+1).key<pairs.get(j).key) {
                Pair tmp = pairs.get(j+1);
                pairs.set(j+1, pairs.get(j));
                pairs.set(j, tmp);
                j--;
            }

            arr.add(new ArrayList<>(pairs));
        }

        return arr;
        
    }
}

File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-3.cpp
class Solution {
   public:
    int findMin(vector<int>& arr) {
        // [3,4,5,6,1,2]
        // l = 0, r = 5
        // m = 2
        // arr[l] = 3
        // arr[r] = 2
        // arr[m] = 5
        // arr[m]>r
        // m = l+1
        // l = 3, r = 5
        // m = 4
        // arr[l] = 6
        // arr[r] = 2
        // arr[m] = 1
        // arr[m] < r
        // return arr[m]

        int l = 0, r = arr.size() - 1;

        while (l < r) {
            int m = l + (r - l) / 2;

            if (arr[m] > arr[r]) {
                l = m + 1;
            } else if (arr[m] < arr[r]) {
                r = m;
            }
        }

        return arr[l];
    }
};
